File: apps/api/app/services/streak.py
"""Série (spec §5.1, contrat parcours §3) — portage de `packages/core/src/streak.ts`.

Dates en « jour local » de l'utilisateur : la série suit son calendrier, pas l'UTC.
- Un jour compte s'il contient au moins une séance terminée (non vide).
- 1 protection offerte tous les 10 jours de série, 2 au maximum, consommée automatiquement.
- Une absence couverte par le gel (« je pars quelques jours ») ne casse pas la série ; le gel ne couvre que les jours
  ≥ `frozen_from` (jour local de la déclaration) et ≤ `frozen_until` : pas de réparation rétroactive.
- Lecture : la série affichée vaut 0 si le dernier jour actif est antérieur à hier et que les jours manqués ne sont
  couverts ni par le gel ni par les protections disponibles.
"""

from dataclasses import dataclass, replace
from datetime import date

MAX_FREEZES = 2
FREEZE_EVERY_DAYS = 10


@dataclass(frozen=True)
class Streak:
    current: int = 0
    longest: int = 0
    last_active_date: date | None = None
    freezes_available: int = 0
    frozen_until: date | None = None
    frozen_from: date | None = None


def _day(d: date) -> int:
    return d.toordinal()


def frozen_days_between(streak: Streak, first: date, last: date) -> int:
    """Jours de [first, last] couverts par le gel (bornés par frozen_from et frozen_until)."""
    if streak.frozen_until is None:
        return 0
    start = _day(first)
    if streak.frozen_from is not None:
        start = max(start, _day(streak.frozen_from))
    end = min(_day(last), _day(streak.frozen_until))
    return max(0, end - start + 1)


def _uncovered(streak: Streak, today: date) -> int:
    """Jours manqués entre le dernier jour actif et `today` (exclus) non couverts par le gel."""
    assert streak.last_active_date is not None  # noqa: S101
    missed = _day(today) - _day(streak.last_active_date) - 1
    if missed <= 0:
        return 0
    first = date.fromordinal(_day(streak.last_active_date) + 1)
    last = date.fromordinal(_day(today) - 1)
    return missed - frozen_days_between(streak, first, last)


def record_activity(streak: Streak, today: date) -> Streak:
    """Enregistre une séance terminée le jour `today`."""
    if streak.last_active_date == today:
        return streak

    freezes = streak.freezes_available
    if streak.last_active_date is None:
        current = 1
    else:
        if _day(today) - _day(streak.last_active_date) <= 0:
            return streak  # horloge revenue en arrière : on ignore
        uncovered = _uncovered(streak, today)
        if uncovered <= 0:
            current = streak.current + 1
        elif uncovered <= freezes:
            freezes -= uncovered
            current = streak.current + 1
        else:
            current = 1

    earned = current // FREEZE_EVERY_DAYS - (current - 1) // FREEZE_EVERY_DAYS
    keep_freeze = streak.frozen_until is not None and _day(streak.frozen_until) >= _day(today)
    return replace(
        streak,
        current=current,
        longest=max(streak.longest, current),
        last_active_date=today,
        freezes_available=min(MAX_FREEZES, freezes + earned),
        frozen_until=streak.frozen_until if keep_freeze else None,
        frozen_from=streak.frozen_from if keep_freeze else None,
    )


def displayed_current(streak: Streak, today: date) -> int:
    """Série affichée le jour `today` (règle de lecture, §3)."""
    if streak.last_active_date is None:
        return 0
    if _day(today) - _day(streak.last_active_date) <= 1:
        return streak.current
    return streak.current if _uncovered(streak, today) <= streak.freezes_available else 0


def freeze(streak: Streak, local_date: date, frozen_until: date | None) -> Streak:
    """Déclaration (ou annulation avec None) du gel « je pars quelques jours »."""
    if frozen_until is None:
        return replace(streak, frozen_until=None, frozen_from=None)
    return replace(streak, frozen_until=frozen_until, frozen_from=local_date)

File: packages/core/src/streak.ts
/**
 * Série (spec §5.1). Dates en « jour local » AAAA-MM-JJ : la série suit le
 * calendrier de l'utilisateur, pas l'UTC.
 *
 * - Un jour compte s'il contient au moins une séance terminée.
 * - 1 protection offerte tous les 10 jours de série, 2 au maximum, consommée automatiquement.
 * - Une absence couverte par `frozenUntil` (vacances déclarées) ne casse pas la série ; le gel ne couvre
 *   que les jours ≥ `frozenFrom` (jour de la déclaration) : pas de réparation rétroactive (contrat phase5 §3).
 */

export interface Streak {
  current: number;
  longest: number;
  lastActiveDate: string | null;
  freezesAvailable: number;
  frozenUntil: string | null;
  /** Jour local de la déclaration du gel ; null = gel ancien sans date (couvre toute l'absence). */
  frozenFrom?: string | null;
}

export const MAX_FREEZES = 2;
export const FREEZE_EVERY_DAYS = 10;

export const emptyStreak = (): Streak => ({
  current: 0,
  longest: 0,
  lastActiveDate: null,
  freezesAvailable: 0,
  frozenUntil: null,
  frozenFrom: null,
});

export function dayNumber(isoDay: string): number {
  const [y, m, d] = isoDay.split("-").map(Number);
  return Math.floor(Date.UTC(y ?? 1970, (m ?? 1) - 1, d ?? 1) / 86_400_000);
}

export function localDay(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, "0");
  const d = String(date.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

/** Enregistre une séance terminée le jour `today`. */
export function recordActivity(streak: Streak, today: string): Streak {
  if (streak.lastActiveDate === today) return streak;

  let current: number;
  let freezes = streak.freezesAvailable;

  if (streak.lastActiveDate === null) {
    current = 1;
  } else {
    const gap = dayNumber(today) - dayNumber(streak.lastActiveDate);
    if (gap <= 0) return streak; // horloge revenue en arrière : on ignore
    const missed = gap - 1;
    const frozenDays = frozenDaysBetween(streak, dayNumber(streak.lastActiveDate) + 1, dayNumber(today) - 1);
    const uncovered = missed - frozenDays;
    if (uncovered <= 0) {
      current = streak.current + 1;
    } else if (uncovered <= freezes) {
      freezes -= uncovered;
      current = streak.current + 1;
    } else {
      current = 1;
    }
  }

  const earned = Math.floor(current / FREEZE_EVERY_DAYS) - Math.floor((current - 1) / FREEZE_EVERY_DAYS);
  return {
    current,
    longest: Math.max(streak.longest, current),
    lastActiveDate: today,
    freezesAvailable: Math.min(MAX_FREEZES, freezes + earned),
    ...stillFrozen(streak, today),
  };
}

function stillFrozen(streak: Streak, today: string): Pick<Streak, "frozenUntil" | "frozenFrom"> {
  const active = streak.frozenUntil !== null && dayNumber(streak.frozenUntil) >= dayNumber(today);
  return active ? { frozenUntil: streak.frozenUntil, frozenFrom: streak.frozenFrom ?? null } : { frozenUntil: null, frozenFrom: null };
}

/** Jours de [from, to] (numéros de jour) couverts par le gel déclaré. */
function frozenDaysBetween(streak: Streak, from: number, to: number): number {
  if (streak.frozenUntil === null || to < from) return 0;
  const start = streak.frozenFrom ? Math.max(from, dayNumber(streak.frozenFrom)) : from;
  const end = Math.min(to, dayNumber(streak.frozenUntil));
  return Math.max(0, end - start + 1);
}

/**
 * Série affichée au jour `today` (lecture, contrat phase5 §3) : 0 si le dernier jour actif est avant
 * hier et que les jours manqués (jusqu'à hier) ne sont couverts ni par le gel ni par les protections.
 */
export function streakAt(streak: Streak, today: string): Streak {
  if (streak.lastActiveDate === null) return streak;
  const gap = dayNumber(today) - dayNumber(streak.lastActiveDate);
  if (gap <= 1) return streak;
  const missed = gap - 1;
  const uncovered = missed - frozenDaysBetween(streak, dayNumber(streak.lastActiveDate) + 1, dayNumber(today) - 1);
  return uncovered <= streak.freezesAvailable ? streak : { ...streak, current: 0 };
}
